Long Parse on folder size results in UI Error app crash

## Issue explanation (write below this line)

while trying to navigate to "/sys/class/power_supply/". batt_verify is a folder, so I don't know why you would parse the name as a long. does java have a try parse or should you just catch the exception? 


## Exception
* __App Name:__ Amaze Dateimanager
* __Package:__ com.amaze.filemanager
* __Version:__ 3.8.2
* __User Action:__ UI Error
* __Request:__ Application crash
* __OS:__ Linux Android 13 - 33
<details><summary><b>Crash log </b></summary><p>

```
java.lang.RuntimeException: An error occurred while executing doInBackground()
	at android.os.AsyncTask$4.done(AsyncTask.java:415)
	at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:381)
	at java.util.concurrent.FutureTask.setException(FutureTask.java:250)
	at java.util.concurrent.FutureTask.run(FutureTask.java:269)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1137)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:637)
	at java.lang.Thread.run(Thread.java:1012)
Caused by: java.lang.NumberFormatException: For input string: "batt_verify"
	at java.lang.Long.parseLong(Long.java:736)
	at java.lang.Long.parseLong(Long.java:861)
	at com.amaze.filemanager.filesystem.files.FileUtils.parseName(FileUtils.java:926)
	at com.amaze.filemanager.filesystem.root.ListFilesCommand.parseStringForHybridFile(ListFilesCommand.kt:222)
	at com.amaze.filemanager.filesystem.root.ListFilesCommand.listFiles(ListFilesCommand.kt:60)
	at com.amaze.filemanager.asynchronous.asynctasks.LoadFilesListTask.doInBackground(LoadFilesListTask.java:265)
	at com.amaze.filemanager.asynchronous.asynctasks.LoadFilesListTask.doInBackground(LoadFilesListTask.java:78)
	at android.os.AsyncTask$3.call(AsyncTask.java:394)
	at java.util.concurrent.FutureTask.run(FutureTask.java:264)
	... 3 more

```
</details>
<hr>

--- a/app/src/main/java/com/amaze/filemanager/filesystem/files/FileUtils.java
+++ b/app/src/main/java/com/amaze/filemanager/filesystem/files/FileUtils.java
@@ -899,7 +899,7 @@ public class FileUtils {
       }
     }
     int p = getColonPosition(array);
-    if (p != -1) {
+    if (p != -1 && (p + 1) != array.length) {
       date = array[p - 1] + " | " + array[p];
       size = array[p - 2];
     } else if (isStat) {
@@ -916,6 +916,10 @@ public class FileUtils {
       int q = getLinkPosition(array);
       for (int i = p + 1; i < q; i++) {
         name.append(" ").append(array[i]);
+      }
+      // Newer *boxes may introduce full path during stat. Trim down to the very last /
+      if (name.lastIndexOf("/") > 0) {
+        name.delete(0, name.lastIndexOf("/") + 1);
       }
       name = new StringBuilder(name.toString().trim());
       for (int i = q + 1; i < array.length; i++) {

--- a/app/src/test/java/com/amaze/filemanager/filesystem/files/FileUtilsTest.kt
+++ b/app/src/test/java/com/amaze/filemanager/filesystem/files/FileUtilsTest.kt
@@ -25,7 +25,9 @@ import android.os.Build.VERSION_CODES.KITKAT
 import android.os.Build.VERSION_CODES.P
 import androidx.test.ext.junit.runners.AndroidJUnit4
 import com.amaze.filemanager.filesystem.files.FileUtils.getPathsInPath
-import org.junit.Assert.*
+import org.junit.Assert.assertArrayEquals
+import org.junit.Assert.assertEquals
+import org.junit.Assert.assertNull
 import org.junit.Test
 import org.junit.runner.RunWith
 import org.robolectric.annotation.Config
@@ -367,13 +369,16 @@ class FileUtilsTest {
      */
     @Test
     fun testParseStringForHybridFileParcelable() {
-        val systemTz = TimeZone.getDefault()
-        TimeZone.setDefault(TimeZone.getTimeZone("UTC"))
-
+        /* ktlint-disable max-line-length */
         // ls
-        val a = "-rwxr-x---   1 root   shell    29431 2009-01-01 08:00 init.rc"
-        val b = "lrw-r--r--   1 root   root        15 2009-01-01 08:00 product -> /system/product"
-        val c = "drwxr-xr-x  17 root   root      4096 1970-05-19 08:40 system"
+        val lsLines = arrayOf(
+            "-rwxr-x---   1 root   shell    29431 2009-01-01 08:00 init.rc",
+            "lrw-r--r--   1 root   root        15 2009-01-01 08:00 product -> /system/product",
+            "drwxr-xr-x  17 root   root      4096 1970-05-19 08:40 system",
+            "-r--r--r-- 1 root root 10 1970-01-13 07:32 cpu_variant:arm",
+            "lrwxrwxrwx  1 root root 0 2022-10-05 15:39 ac -> ../../devices/platform/GFSH0001:00/power_supply/ac",
+            "lrwxrwxrwx   1 root root 0 2022-10-05 00:16 usb -> ../../devices/platform/soc/c440000.qcom,spmi/spmi-0/spmi0-02/c440000.qcom,spmi:qcom,pm8150b@2:qcom,qpnp-smb5/power_supply/usb"
+        )
 
         // stat with old toybox or busybox
         // val a1 = "-rwxr-x--- 1 shell root 512 2009-01-01 08:00:00.000000000 `init.rc'"
@@ -381,29 +386,33 @@ class FileUtilsTest {
         // val c1 = "drwxr-xr-x 17 root root 512 1970-05-19 08:40:27.269999949 `system'"
 
         // stat with new toybox
-        val a2 = "-rwxr-x--- 1 shell root 512 1230796800 `init.rc'"
-        val b2 = "lrw-r--r-- 1 root root 512 1230796800 `product' -> `/system/product'"
-        val c2 = "drwxr-xr-x 17 root root 512 11922027 `system'"
+        val statLines = arrayOf(
+            "-rwxr-x--- 1 shell root 512 1230796800 `init.rc'",
+            "lrw-r--r-- 1 root root 512 1230796800 `product' -> `/system/product'",
+            "drwxr-xr-x 17 root root 512 11922027 `system'",
+            "-r--r--r-- 1 root root 512 1035141 `cpu_variant:arm'",
+            "lrwxrwxrwx 1 root root 512 1664955558 /sys/class/power_supply/ac -> '../../devices/platform/GFSH0001:00/power_supply/ac'",
+            "lrwxrwxrwx 1 root root 512 1664956626 /sys/class/power_supply/usb -> '../../devices/platform/soc/c440000.qcom,spmi/spmi-0/spmi0-02/c440000.qcom,spmi:qcom,pm8150b@2:qcom,qpnp-smb5/power_supply/usb'"
+        )
+        /* ktlint-enable max-line-length */
 
-        var result1 = FileUtils.parseName(a, false)
-        var result2 = FileUtils.parseName(a2.replace("('|`)".toRegex(), ""), true)
-        assertEquals(result1.date, result2.date)
-        assertEquals(result1.name, result2.name)
-        assertEquals(result1.path, result2.path)
+        val systemTz = TimeZone.getDefault()
+        TimeZone.setDefault(TimeZone.getTimeZone("UTC"))
 
-        result1 = FileUtils.parseName(b, false)
-        result2 = FileUtils.parseName(b2.replace("('|`)".toRegex(), ""), true)
-        assertEquals(result1.date, result2.date)
-        assertEquals(result1.name, result2.name)
-        assertEquals(result1.path, result2.path)
-        assertEquals(result1.link, result2.link)
-
-        result1 = FileUtils.parseName(c, false)
-        result2 = FileUtils.parseName(c2.replace("('|`)".toRegex(), ""), true)
-        // if using stat, seconds will also be available, so they won't be equal
-        assertNotEquals(result1.date, result2.date)
-        assertEquals(result1.name, result2.name)
-        assertEquals(result1.path, result2.path)
+        lsLines.forEachIndexed { index: Int, s: String ->
+            val result1 = FileUtils.parseName(s, false)
+            val result2 = FileUtils.parseName(statLines[index].replace("('|`)".toRegex(), ""), true)
+            assertEquals(
+                "Parse error at index $index.\n lsLines=[$s]\n statLines=[${statLines[index]}]\n",
+                result1.name,
+                result2.name
+            )
+            assertEquals(
+                "Parse error at index $index.\n lsLines=[$s]\n statLines=[${statLines[index]}]\n",
+                result1.path,
+                result2.path
+            )
+        }
 
         TimeZone.setDefault(systemTz)
     }

